### PR TITLE
feat(utils): extract_if backport

### PR DESCRIPTION
**Pull request by sinui0**
This PR replaces our existing `FilterDrain`, backporting the std implementation of `extract_if` until it reaches stabilization.

Source code is directly copied (minus renaming) including the tests.

### CODE REVIEW

File: utils/src/filter_drain.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+/// An iterator which uses a closure to determine if an element should be removed.
+#[derive(Debug)]
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+pub struct FilterDrainIter<'a, T, F>
+where
+    F: FnMut(&mut T) -> bool,
+{
+    pub(super) vec: &'a mut Vec<T>,
```
Comment (th4s) [inline utils/src/filter_drain.rs:54]: I think we do not need `pub(super)` here, because we create `FilterDrainIter` in the same module.

**Review by th4s — APPROVED** [review]
:+1:

File: utils/src/filter_drain.rs (hunk 1 above)
Comment (sinui0) [reply utils/src/filter_drain.rs:54]: I just copy pasted exactly from the std source, it should be fine

Comment (th4s) [reply utils/src/filter_drain.rs:54]: In the end it doesn't matter, but you use a trait in the same module to instantiate it, while the standard library does the instantiation in the super module.

**Pull request merged**